### ISSUE TITLE
HTTP Integration with endpoint that never answers HTTP 200 or never listen at all is making ChirpStack to fail Class A sensors downlinks

### ISSUE THREAD

**Issue by mouchh** [chirpstack/chirpstack#364]
Our ChirpStack started to behave strange:

- We couldn’t find anymore Link Metrics when browsing any Device dashboard
![image](https://github.com/chirpstack/chirpstack/assets/3480808/3edb2123-71cd-421e-9920-186f762f6791)

- Our sensors stopped using ADR (Adaptive Data Rate) although they sent message with ADR flag to `True` - they switched back to DR0 one after the other
![image](https://github.com/chirpstack/chirpstack/assets/3480808/aed059f7-dc77-4f3a-9227-93f400034a0e)

- ChirpStack was not able to send downlink anymore (apart from the initial JoinAccept)
  - Either via the Enqueue menu - although uplink were sent (we use Class A devices), the queue remained untouched
  ![image](https://github.com/chirpstack/chirpstack/assets/3480808/b3883474-2def-4c2a-8631-bb6887c8f4a8)

  - Or just sending Confirmation Downlink for the sensors requiring it (see below how our ChirpStack is never sending the requested confirmation, making our sensor unable to close its Join procedure)
  ![image](https://github.com/chirpstack/chirpstack/assets/3480808/ef693931-1707-436d-b60c-2110c8ebce39)

Apart from these, we could still receive going data, and we could even Join new sensor on the network (unless their Join procedure requires another ACK via a Confirmed downlink).

We realized (after days of analysis - the logs were silent on anything going wrong here) that these observations were directly linked to the way we configured our HTTP Integration (per Application profile in ChirpStack).

## Situation 1: NodeRed HTTP Server without HTTP 200 response -> ChirpStack hanging for HTTP ack/answer

So - we lately integrated a HTTP Integration endpoint; which is in fact a NodeRed instance listening for HTTP payloads.
But within NodeRed, we didnt create a "HTTP Response" node that would send back with HTTP 200 flag any for any client hooking into.
See the "http" reponse node, not linked with the input node (listening for HTTP payload)

![image](https://github.com/chirpstack/chirpstack/assets/3480808/57791944-4091-496e-865b-e7bb982e8a4c)

Here is the Wireshark capture (from the NodeRed host, the one with 155.X.X.X IP)
![image](https://github.com/chirpstack/chirpstack/assets/3480808/171ba6e0-8a07-420b-bee5-d74d402078a4)

-	TCP ACK for HTTP payload event=up received -> packet 99
-	No HTTP ACK (HTTP 200) for HTTP payload event=up received!
-	Nothing is then exchanged for 5 seconds (time for packet 124 minus time for packet 99 and ChirpStack finally closes the socket -> packet 124 (be careful as 2 socket streams are mixed in this capture)

And the same behavior happens for the HTTP payload event=log

## Situation 2: NodeRed HTTP Server with HTTP 200 response -> ChirpStack not hanging

If instead you configure the HTTP response in your NodeRed instance (see below); ChirpStack gets back to a normal ways of working (all previous observations vanished)

![image](https://github.com/chirpstack/chirpstack/assets/3480808/72482798-7c07-4cbb-9ddc-35dfda833196)

Here is the Wireshark capture (from the NodeRed host, the one with 155.X.X.X IP)
![image](https://github.com/chirpstack/chirpstack/assets/3480808/822c0673-2d8b-4d63-b030-56f4c2ffa80e)

-	TCP ACK for HTTP payload event=up received -> packet 294
-	HTTP ACK (HTTP 200) for HTTP payload event=up received -> packet 295

This happened in milliseconds (and there is almost no impact due to network latency, both machines are very close in the network).
Right after, ChirpStack is initiating the TCP socket closure (packet 297). All in all, it took 10ms to send the HTTP message, have the HTTP 200 and closing the socket.

The same behavior reoccurs right after for HTTP payload event=log, as well sent by the ChirpStack HTTP integration.

## Situation 3: Complete timeout for HTTP Integration (misconfig for ex) -> ChirpStack hanging for TCP ACK

This is the final test we ran; misconfiguring a HTTP endpoint that isnt one to see how ChirpStack is behaving.

It is actually timing out (obviously), but indeed in the same time that ChirpStack is waiting for its TCP handshake to complete, it is not looking at any downlink events that should have occurred in the meantime (for Class A sensors particularly).

I think this past issue described already this behavior (https://github.com/chirpstack/chirpstack/issues/53) - which was not detailed enough for proper analysis and resolution.

# Conclusion

This behavior (HTTP integration hanging ChirpStack waiting for TCP ACK or HTTP ACK) happens per Application profile - if no HTTP Integration are configured for Application B, the configured for Application A will not impact ChirpStack downlinks for Application B.
With these tests, I believe that ChirpStack integrations (at least HTTP ones) have an impact on the actual ChirpStack downlink functionalities.   
I am thinking about Integrations as Hooks (same than receiving a mail whenever someone commit to my Git repo); as such, it sounds very strange that whatever happens on the Integration (time out, or no HTTP 200 response for example) could actually impact the way primary app is working. You still want your commit to exists even though mail notification couldnt have been sent. And you still want ChirpStack to schedule proper downlink even though a hooked HTTP endpoint is not responding as expected.

If you agree with this last statement, I think we have a bug here; as from my previous tests, Integrations are having a direct impact on ChirpStack inherent communication with sensors.  
**It feels like during the 5 seconds where ChirpStack is waiting for HTTP ACK (situation 1) or TCP ACK (situation 3), it is not able to schedule downlink communication. These hanging seconds appear at the worst moment for Class A sensors = at the moment they actually have their RX Window ready for downlink!**

We probably dont want any overlap/ressource vampirization between 
- communication with `Integrations` and 
- communication with `Devices`

I could "solve" this issue by making sure we feedback HTTP 200 response; to release ChirpStack as soon as possible, and making it available for comming RX windows but it dont feels like a sustainable solution. 

Happy to hear from you!

**Comment by brocaar** [comment]
I'm not sure if I can reproduce your issue:

![image](https://github.com/chirpstack/chirpstack/assets/165497/9f8f27ce-4604-4072-866b-8e33247b0c89)

NodeRED flow without HTTP response.

```
2024-02-06T15:05:42.697388Z  INFO chirpstack::integration::http: Posting event event=up url=http://10.0.0.243:1880/test
2024-02-06T15:05:42.699246Z  INFO chirpstack::integration::mqtt: Publishing event topic=application/00000000-0000-0000-0000-000000000001/device/0101010101010101/event/up
2024-02-06T15:05:42.702222Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}: chirpstack::storage::device_session: Device-session saved dev_eui=0101010101010101 dev_addr=016ec9ab
2024-02-06T15:05:42.702916Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}: chirpstack::storage::metrics: Metrics saved name=device:0101010101010101 aggregation=HOUR
2024-02-06T15:05:42.703320Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}: chirpstack::storage::metrics: Metrics saved name=device:0101010101010101 aggregation=DAY
2024-02-06T15:05:42.703726Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}: chirpstack::storage::metrics: Metrics saved name=device:0101010101010101 aggregation=MONTH
2024-02-06T15:05:42.810307Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}:data_down{downlink_id=1783800741}: chirpstack::storage::mac_command: Pending mac-command block set dev_eui=0101010101010101 cid=LinkADRReq
2024-02-06T15:05:42.811059Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}:data_down{downlink_id=1783800741}: chirpstack::storage::downlink_frame: Downlink-frame saved downlink_id=1783800741
2024-02-06T15:05:42.811929Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}:data_down{downlink_id=1783800741}: chirpstack::storage::device_session: Device-session saved dev_eui=0101010101010101 dev_addr=016ec9ab
2024-02-06T15:05:42.812112Z  INFO up{deduplication_id=5541be9b-7205-405e-b0be-27b0c6fbde99}:data_up{dev_eui="0101010101010101"}:data_down{downlink_id=1783800741}: chirpstack::gateway::backend::mqtt: Sending downlink frame region_id=eu868 gateway_id=0016c001ff1a0a5f topic=eu868/gateway/0016c001ff1a0a5f/command/down json=false
2024-02-06T15:05:42.816979Z  INFO chirpstack::gateway::backend::mqtt: Message received from gateway region_id="eu868" topic=eu868/gateway/0016c001ff1a0a5f/event/ack qos=AtMostOnce json=false
2024-02-06T15:05:42.818332Z  INFO tx_ack{downlink_id=1783800741}: chirpstack::storage::device_session: Device-session saved dev_eui=0101010101010101 dev_addr=016ec9ab
2024-02-06T15:05:42.818377Z  INFO tx_ack{downlink_id=1783800741}: chirpstack::downlink::tx_ack: Log downlink-frame for gateway gateway_id=0016c001ff1a0a5f
2024-02-06T15:05:42.818888Z  INFO tx_ack{downlink_id=1783800741}: chirpstack::downlink::tx_ack: Log downlink-frame for device device_eui=0101010101010101
2024-02-06T15:05:47.698348Z  WARN chirpstack::integration::http: Posting event failed event=up url=http://10.0.0.243:1880/test error=error sending request for url (http://10.0.0.243:1880/test?event=up): operation timed out
```

* I sent a confirmed uplink
* This was sent to the NodeRED endpoint
* Meanwhile the downlink containing the Ack was sent to the gateway
* The gateway returned a TX Ack
* After that, the HTTP integration post timed out

As you can see here, we do not wait for the HTTP request to complete (it is dispatched as a tokio task):
https://github.com/chirpstack/chirpstack/blob/master/chirpstack/src/integration/mod.rs#L196

**Comment by mouchh** [comment]
Interesting, thanks for the note.  
I see you pushed a fix (https://github.com/chirpstack/chirpstack/commit/aa9923a60b38bd9a9baaf0522d9dba4893743904) for these functions back in the v4.1.2... decoupling slow integrations from downlinks!

I bet my production instance is actually running an older version than this one. Will keep you posted as soon as possible

**Comment by mouchh** [comment]
> Interesting, thanks for the note. I see you pushed a fix ([aa9923a](https://github.com/chirpstack/chirpstack/commit/aa9923a60b38bd9a9baaf0522d9dba4893743904)) for these functions back in the v4.1.2... decoupling slow integrations from downlinks!
> 
> I bet my production instance is actually running an older version than this one. Will keep you posted as soon as possible

Confirmed. My production instance is version 4.0.2 😢
This bug has been solved already!
